### PR TITLE
Add PartitionedTable  transform() and partitioned_transform() methods

### LINKED ISSUE #2549: Add PartitionedTable transform methods to the Python wrapper

### PR DESCRIPTION
**Pull request by jmao-denver**
Fixes #2549

### CODE REVIEW

File: Integrations/src/main/java/io/deephaven/integrations/python/PythonBiFunction.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        /**
+         * Creates a {@link PythonBinaryOperator} which calls a Python function.
+         *
+         * @param pyCallable the python object providing the function - must either be callable or have an `apply`
+         *        attribute which is callable.
+         * @param classOut the specific java class to interpret the return for the method. This can be one of String,
+         *        double, float, long, int, short, byte, or boolean; or in the case of Python wrapper objects, PyObject,
+         *        such objects then can be unwrapped to be used inside Java.
+         */
+        public PythonBinaryOperator(PyObject pyCallable, Class classOut) {
```
Comment (rcaudy) [inline Integrations/src/main/java/io/deephaven/integrations/python/PythonBiFunction.java:58]:
```suggestion
        public PythonBinaryOperator(PyObject pyCallable, Class<T> classOut) {
```

File: Integrations/src/main/java/io/deephaven/integrations/python/PythonBiFunction.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+/**
+ * A {@link BiFunction} implementation which calls a Python callable.
+ * 
+ * @param <T> input argument class
+ * @param <U> input argument class
+ */
+@ScriptApi
+public class PythonBiFunction<T, U, R> implements BiFunction<T, U, R> {
+    private final PyObject pyCallable;
+    private final Class classOut;
```
Comment (rcaudy) [inline Integrations/src/main/java/io/deephaven/integrations/python/PythonBiFunction.java:23]:
```suggestion
    private final Class<R> classOut;
```

File: Integrations/src/main/java/io/deephaven/integrations/python/PythonBiFunction.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        this.pyCallable = pyApplyFunc(pyCallable);
+        this.classOut = classOut;
+    }
+
+    @Override
+    public R apply(T t, U u) {
+        PyObject wrapped = PythonObjectWrapper.wrap(t);
+        PyObject wrappedOther = PythonObjectWrapper.wrap(u);
+        PyObject out = pyCallable.call("__call__", wrapped, wrappedOther);
+        return (R) PythonValueGetter.getValue(out, classOut);
```
Comment (rcaudy) [inline Integrations/src/main/java/io/deephaven/integrations/python/PythonBiFunction.java:44]:
```suggestion
        return PythonValueGetter.getValue(out, classOut);
```

File: Integrations/src/main/java/io/deephaven/integrations/python/PythonFunction.java [hunk 4]
```diff
@@ -16,140 +17,46 @@
  * @param <T> input argument class
  */
 @ScriptApi
-public class PythonFunction<T> implements Function<T, Object> {
+public class PythonFunction<T, R> implements Function<T, R> {
     private final PyObject pyCallable;
-    private final Function<PyObject, Object> getter;
+    private final Class classOut;
```
Comment (rcaudy) [inline Integrations/src/main/java/io/deephaven/integrations/python/PythonFunction.java:22]:
```suggestion
    private final Class<R> classOut;
```

File: Integrations/src/main/java/io/deephaven/integrations/python/PythonFunction.java [hunk 5]
```diff
@@ -16,140 +17,46 @@
...
+        /**
+         * Creates a {@link UnaryOperator} which calls a Python function.
+         *
+         * @param pyCallable the python object providing the function - must either be callable or have an `apply`
+         *        attribute which is callable.
+         * @param classOut the specific java class to interpret the return for the method. This can be one of String,
+         *        double, float, long, int, short, byte, or boolean; or in the case of Python wrapper objects, PyObject,
+         *        such objects then can be unwrapped to be used inside Java.
+         */
+        public PythonUnaryOperator(PyObject pyCallable, Class classOut) {
```
Comment (rcaudy) [inline Integrations/src/main/java/io/deephaven/integrations/python/PythonFunction.java:58]:
```suggestion
        public PythonUnaryOperator(PyObject pyCallable, Class<T> classOut) {
```

File: Integrations/src/main/java/io/deephaven/integrations/python/PythonFunction.java [hunk 6]
```diff
@@ -16,140 +17,46 @@
...
-        @Override
-        public Object apply(PyObject valueIn) {
-            if (valueIn == null) {
-                return null;
-            }
-            return valueIn.getBooleanValue();
+    public R apply(T t) {
+        PyObject wrapped = PythonObjectWrapper.wrap(t);
+        PyObject out = pyCallable.call("__call__", wrapped);
+        return (R) PythonValueGetter.getValue(out, classOut);
```
Comment (rcaudy) [inline Integrations/src/main/java/io/deephaven/integrations/python/PythonFunction.java:44]:
```suggestion
        return PythonValueGetter.getValue(out, classOut);
```

File: Integrations/src/main/java/io/deephaven/integrations/python/PythonObjectWrapper.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    private static final PyModule PY_WRAPPER_MODULE = PyModule.importModule("deephaven._wrapper");
+
+    /**
+     * Unwrap a Python object to return the wrapped Java object.
+     * 
+     * @param t An instance of {@link PyObject}.
+     * @return The wrapped Java object.
+     */
+    public static Object unwrap(PyObject t) {
+        return PY_WRAPPER_MODULE.call("unwrap", t).getObjectValue();
```
Comment (rcaudy) [inline Integrations/src/main/java/io/deephaven/integrations/python/PythonObjectWrapper.java:23]:
```suggestion
        //noinspection ConstantConditions
        return PY_WRAPPER_MODULE.call("unwrap", t).getObjectValue();
```

File: Integrations/src/main/java/io/deephaven/integrations/python/PythonFunction.java [hunk 8]
```diff
@@ -16,140 +17,46 @@
...
+    public static class PythonUnaryOperator<T> extends PythonFunction<T, T> implements UnaryOperator<T> {
+
+        /**
+         * Creates a {@link UnaryOperator} which calls a Python function.
+         *
+         * @param pyCallable the python object providing the function - must either be callable or have an `apply`
+         *        attribute which is callable.
+         * @param classOut the specific java class to interpret the return for the method. This can be one of String,
+         *        double, float, long, int, short, byte, or boolean; or in the case of Python wrapper objects, PyObject,
+         *        such objects then can be unwrapped to be used inside Java.
```
Comment (rcaudy) [inline Integrations/src/main/java/io/deephaven/integrations/python/PythonFunction.java:56]:
```suggestion
         * @param pyCallable The python object providing the function - must either be callable or have an
         *        {@code apply} attribute which is callable.
         * @param classOut The specific Java class expected to be returned by the {@link #apply(Object)} method. This
         *        should be the result of converting or unwrapping the output of {@code pyCallable}.
```

File: Integrations/src/main/java/io/deephaven/integrations/python/PythonFunction.java [hunk 9]
```diff
@@ -16,140 +17,46 @@
  * @param <T> input argument class
  */
 @ScriptApi
```
Comment (rcaudy) [inline Integrations/src/main/java/io/deephaven/integrations/python/PythonFunction.java:19]:
```suggestion
@SuppressWarnings("unused")
@ScriptApi
```

File: Integrations/src/main/java/io/deephaven/integrations/python/PythonBiFunction.java [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+import static io.deephaven.integrations.python.PythonUtils.pyApplyFunc;
+
+/**
+ * A {@link BiFunction} implementation which calls a Python callable.
+ * 
+ * @param <T> input argument class
+ * @param <U> input argument class
+ */
+@ScriptApi
```
Comment (rcaudy) [inline Integrations/src/main/java/io/deephaven/integrations/python/PythonBiFunction.java:20]:
```suggestion
@SuppressWarnings("unused")
@ScriptApi
```

File: Integrations/src/main/java/io/deephaven/integrations/python/PythonBiFunction.java [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+
+    public static class PythonBinaryOperator<T> extends PythonBiFunction<T, T, T> implements BinaryOperator<T> {
+        /**
+         * Creates a {@link PythonBinaryOperator} which calls a Python function.
+         *
+         * @param pyCallable the python object providing the function - must either be callable or have an `apply`
+         *        attribute which is callable.
+         * @param classOut the specific java class to interpret the return for the method. This can be one of String,
+         *        double, float, long, int, short, byte, or boolean; or in the case of Python wrapper objects, PyObject,
+         *        such objects then can be unwrapped to be used inside Java.
```
Comment (rcaudy) [inline Integrations/src/main/java/io/deephaven/integrations/python/PythonBiFunction.java:56]:
```suggestion
         * @param pyCallable The python object providing the function - must either be callable or have an
         *        {@code apply} attribute which is callable.
         * @param classOut The specific Java class expected to be returned by the {@link #apply(Object, Object)} method.
         *        This should be the result of converting or unwrapping the output of {@code pyCallable}.
```

File: Integrations/src/main/java/io/deephaven/integrations/python/PythonBiFunction.java [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+    private final Class classOut;
+
+    /**
+     * Creates a {@link BiFunction} which calls a Python function.
+     *
+     * @param pyCallable the python object providing the function - must either be callable or have an `apply` attribute
+     *        which is callable.
+     * @param classOut the specific java class to interpret the return for the method. This can be one of String,
+     *        double, float, long, int, short, byte, or boolean; or in the case of Python wrapper objects, PyObject,
+     *        such objects then can be unwrapped to be used inside Java.
```
Comment (rcaudy) [inline Integrations/src/main/java/io/deephaven/integrations/python/PythonBiFunction.java:32]:
```suggestion
     * @param pyCallable The python object providing the function - must either be callable or have an {@code apply}
     *        attribute which is callable.
     * @param classOut The specific Java class expected to be returned by the {@link #apply(Object, Object)} method.
     *        This should be the result of converting or unwrapping the output of {@code pyCallable}.
```

File: Integrations/src/main/java/io/deephaven/integrations/python/PythonFunction.java [hunk 13]
```diff
@@ -16,140 +17,46 @@
...
+public class PythonFunction<T, R> implements Function<T, R> {
     private final PyObject pyCallable;
-    private final Function<PyObject, Object> getter;
+    private final Class classOut;
 
     /**
      * Creates a {@link Function} which calls a Python function.
      *
-     * @param pyObjectIn the python object providing the function - must either be callable or have an `apply` attribute
+     * @param pyCallable the python object providing the function - must either be callable or have an `apply` attribute
```
Comment (rcaudy) [inline Integrations/src/main/java/io/deephaven/integrations/python/PythonFunction.java:27]:
```
     * @param pyCallable The python object providing the function - must either be callable or have an {@code apply}
     *        attribute which is callable.
     * @param classOut The specific Java class expected to be returned by the {@link #apply(Object)} method. This
     *        should be the result of converting or unwrapping the output of {@code pyCallable}.
```

File: py/server/deephaven/table.py [hunk 14]
```diff
@@ -1451,3 +1451,47 @@ def get_constituent(self, key_values: Union[Any, Sequence[Any]]) -> Optional[Tab
...
+            func (Callable[[Table, Table], Table]: a function which takes two Tables as input and returns a new Table
+
+        Returns:
+            a PartitionedTable
+
+        Raises:
+            DHError
+        """
+        try:
+            j_pt = self.j_partitioned_table.partitionedTransform(other.j_partitioned_table, j_binary_operator(func, dtypes.PyObject))
```
Comment (rcaudy) [inline py/server/deephaven/table.py:1494]:
```suggestion
            j_pt = self.j_partitioned_table.partitionedTransform(other.j_partitioned_table, j_binary_operator(func, dtypes.from_jtype(Table.j_object_type)))
```

File: py/server/deephaven/table.py [hunk 15]
```diff
@@ -1451,3 +1451,47 @@ def get_constituent(self, key_values: Union[Any, Sequence[Any]]) -> Optional[Tab
...
+            func (Callable[[Table], Table]: a function which takes a Table as input and returns a new Table
+
+        Returns:
+            a PartitionedTable
+
+        Raises:
+            DHError
+        """
+        try:
+            j_pt = self.j_partitioned_table.transform(j_unary_operator(func, dtypes.PyObject))
```
Comment (rcaudy) [inline py/server/deephaven/table.py:1470]:
```suggestion
            j_pt = self.j_partitioned_table.transform(j_unary_operator(func, dtypes.from_jtype(Table.j_object_type)))
```

File: Integrations/src/main/java/io/deephaven/integrations/python/PythonObjectWrapper.java [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+
+    /**
+     * Wrap a raw Java object with a Python wrapper if one exists for its type, otherwise return a JPY mapped Python
+     * object.
+     * 
+     * @param t A Java object.
+     * @return A {@link PyObject} instance representing the Python wrapper object.
+     */
+    public static PyObject wrap(Object t) {
+        return PY_WRAPPER_MODULE.call("wrap_j_object", t);
```
Comment (rcaudy) [inline Integrations/src/main/java/io/deephaven/integrations/python/PythonObjectWrapper.java:35]:
```suggestion
        //noinspection ConstantConditions
        return PY_WRAPPER_MODULE.call("wrap_j_object", t);
```

**Review by rcaudy — APPROVED** [review]

File: py/server/deephaven/table.py [hunk 17]
```diff
@@ -1451,3 +1451,49 @@ def get_constituent(self, key_values: Union[Any, Sequence[Any]]) -> Optional[Tab
...
+            return PartitionedTable(j_partitioned_table=j_pt)
+        except Exception as e:
+            raise DHError(e, "failed to transform the PartitionedTable.") from e
+
+    def partitioned_transform(self, other: PartitionedTable, func: Callable[[Table, Table], Table]) -> PartitionedTable:
+        """Join the underlying partitioned Tables from this PartitionedTable and other on the key columns, then apply
+        the provided function to all pairs of constituent Tables with the same keys in order to produce a new
+        PartitionedTable with the results as its constituents, with the same data for all other columns in the
+        underlying partitioned Table from this. Note that if the Tables underlying this PartitionedTable or other
+        change, a corresponding change will propagate to the result.
```
Comment (chipkent) [inline py/server/deephaven/table.py:1481]: """Join this and the other PartitionedTables on the key columns and applies the provided function to all pairs
        of constituent tables with the same keys.  The resulting tables are used to produce a new PartitionedTable 
        with the results as its constituents and with the same data as this PartitionedTable for all other columns.

        Note that if the Tables underlying this PartitionedTable or other change, a corresponding change will 
        propagate to the result.

**Pull request merged**